modules/Core.js

```diff
@@ -113,17 +120,57 @@ Route.propTypes = {
 /**
  * Low-level API for rendering the props provided to a <Route> element.
  */
-Route.render = ({ component, render, children, ...props }) => (
-  component ? ( // component prop gets first priority, only called if there's a match
-    props.match ? React.createElement(component, props) : null
-  ) : render ? ( // render prop is next, only called if there's a match
-    props.match ? render(props) : null
-  ) : children ? ( // children come last, always called
-    typeof children === 'function' ? children(props) : React.Children.only(children)
-  ) : (
-    null
-  )
-)
+Route.render = ({ component, render, children, ...props }) => {
+  let element = null
+  if (component) {
+    element = props.match ? React.createElement(component, props) : null
+  } else if (render) {
+    element = props.match ? render(props) : null
+  } else if (children) {
+    element = typeof children === 'function' ? children(props) : React.Children.only(children)
+  }
+
+  return element === null ? null : React.createElement(RoutingProvider, {
+    match: props.match,
+    history: props.history,
+    children: element
+  })
+}
+
+class RoutingProvider extends React.Component {
+  
+  static propTypes = {
+    match: PropTypes.object,
+    history: PropTypes.object
+  }
+
+  static childContextTypes = {
+    router: PropTypes.shape({
+      history: PropTypes.shape({
+        listen: PropTypes.func.isRequired
+      }),
+      match: PropTypes.shape({
+        getMatch: PropTypes.func.isRequired
+      })
+    })
+  }
+
+  getChildContext() {
+    return {
+      router: {
+        history: this.props.history,
+        match: {
+          getMatch: () => this.props.match
+        }
+      }
+    }
+  }
+
+  render() {
+    return React.Children.only(this.props.children)
```

Comment:
This will `throw` when we render a `<Router>` with no `children`. I'd rather just silently render nothing in that scenario.

```js
children ? React.Children.only(children) : null
```